File: example-scala-learn/src/spark/WordCountDemo.scala
package spark

import org.apache.spark.{SparkConf, SparkContext}

/**
  * @description
  * @author zhangwu
  * @date 2017-11-06-21:21
  * @version 1.0.0
  */
object WordCountDemo {
  def main(args: Array[String]) {
    val inputFile="README.md"
    val outputFile="./countWord"
    // 创建一个Scala版本的Spark Context
    val conf = new SparkConf().setAppName("wordCount").setMaster("local")
    val sc = new SparkContext(conf)
    // 读取我们的输入数据
    val input = sc.textFile(inputFile)
    // 把它切分成一个个单词
    val words = input.flatMap(line => line.split(" "))
    // 转换为键值对并计数
    val counts = words.map(word => (word, 1)).reduceByKey{case (x, y) => x + y}
    // 将统计出来的单词总数存入一个文本文件，引发求值
    counts.saveAsTextFile(outputFile)
  }
}

File: example-spark-learn/src/main/java/com/undergrowth/WordCountDemo.java
package com.undergrowth;

import com.undergrowth.spark.util.SparkCommonUtil;
import java.util.Arrays;
import java.util.Iterator;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import scala.Tuple2;

/**
 * @author zhangwu
 * @version 1.0.0
 * @description 单词计数
 * @date 2017-11-06-21:01
 */
public class WordCountDemo {

  public static void main(String[] args) {
    SparkConf conf = SparkCommonUtil.getSparkConf("HelloWorld", "local");
    // setSparkHome指向安装spark的地址，视环境而定
    JavaSparkContext sc = new JavaSparkContext(conf);

    JavaRDD<String> data = sc.textFile("README.md");
    // 切分为单词
    JavaRDD<String> words = data.flatMap(
        new FlatMapFunction<String, String>() {
          public Iterator<String> call(String x) {
            return Arrays.asList(x.split(" ")).iterator();
          }
        });
// 转换为键值对并计数
    JavaPairRDD<String, Integer> counts = words.mapToPair(
        new PairFunction<String, String, Integer>() {
          public Tuple2<String, Integer> call(String x) {
            return new Tuple2(x, 1);
          }
        }).reduceByKey(new Function2<Integer, Integer, Integer>() {
      public Integer call(Integer x, Integer y) {
        return x + y;
      }
    });
// 将统计出来的单词总数存入一个文本文件，引发求值
    System.out.println(counts.count());
  }
}
